pkg/client/client_test.go

```diff
@@ -59,133 +66,248 @@ func TestApplySingle(t *testing.T) {
+				})
+
+				t.Run("it should not update when using same object", func(t *testing.T) {
+					// given
+					cl, _ := newClient(t, s)
+					_, err := cl.CreateOrUpdateObject(defaultService.DeepCopyObject(), true, nil)
+					require.NoError(t, err)
+
+					// when
+					createdOrChanged, err := cl.CreateOrUpdateObject(defaultService.DeepCopyObject(), false, nil)
+
+					// then
+					require.NoError(t, err)
+					assert.False(t, createdOrChanged)
+				})
+
+				t.Run("when object is missing, it should create it", func(t *testing.T) {
+					// given
+					cl, cli := newClient(t, s)
+
+					// when
+					createdOrChanged, err := cl.CreateOrUpdateObject(modifiedService.DeepCopyObject(), false, &appsv1.Deployment{})
+
+					// then
+					require.NoError(t, err)
+					assert.True(t, createdOrChanged)
+					service := &corev1.Service{}
+					err = cli.Get(context.TODO(), namespacedName, service)
+					require.NoError(t, err)
+					assert.Equal(t, "all-services", service.Spec.Selector["run"])
+					assert.NotEmpty(t, service.OwnerReferences)
+				})
+			})
+
+			t.Run("when object cannot be retrieved because of any error, then it should fail", func(t *testing.T) {
+				// given
+				cl, cli := newClient(t, s)
+				cli.MockGet = func(ctx context.Context, key client.ObjectKey, obj runtime.Object) error {
+					return fmt.Errorf("unable to get")
+				}
+
+				// when
+				createdOrChanged, err := cl.CreateOrUpdateObject(modifiedService.DeepCopyObject(), false, nil)
+
+				// then
+				require.Error(t, err)
+				assert.False(t, createdOrChanged)
+				assert.Contains(t, err.Error(), "unable to get the resource")
+			})
 		})
 
-		t.Run("when using forceUpdate=true, it should update when specs are different", func(t *testing.T) {
-			// given
-			cl, _ := newClient(t, s)
-			obj := defaultService.DeepCopy()
-			_, err := cl.CreateOrUpdateObject(obj, true, nil)
-			require.NoError(t, err)
-			originalGeneration := obj.GetGeneration()
-
-			// when updating with the modified obj
-			modifiedObj := modifiedService.DeepCopy()
-			modifiedObj.ObjectMeta.Generation = obj.GetGeneration()
-			createdOrChanged, err := cl.CreateOrUpdateObject(modifiedObj, true, nil)
-
-			// then
-			require.NoError(t, err)
-			assert.True(t, createdOrChanged) // resource was updated on the server, so returned value if `true`
-			updateGeneration := modifiedObj.GetGeneration()
-			assert.Equal(t, originalGeneration+1, updateGeneration)
+		t.Run("as unstructured objects", func(t *testing.T) {
+
+			// only testing the specific cases of Services where an existing version exists, with a `spec.clusterIP` set
+			// and the updated version has no value for this field
+
+			t.Run("when using forceUpdate=true", func(t *testing.T) {
+
+				t.Run("it should not update when specs are same except ClusterIP", func(t *testing.T) {
+					// given
+					cl, _ := newClient(t, s)
+					// convert to unstructured
+					obj, err := toUnstructured(defaultService.DeepCopy())
```

Comment:
you could use 
```
runtime.DefaultUnstructuredConverter.ToUnstructured(defaultService.DeepCopy())
```

Reply:
sadly, `runtime.DefaultUnstructuredConverter.ToUnstructured(defaultService.DeepCopy())` returns `map[string]interface{}` and I need an `*unstructured.Unstructured` (which implements `runtime.Object`)

Reply:
ah, right... you would need to do:
```
converted, err := runtime.DefaultUnstructuredConverter.ToUnstructured(defaultService.DeepCopy())
require.NoError(t, err)
obj := &unstructured.Unstructured{
	Object: converted,
}
```

Reply:
are you asking me to apply this change in the `toUnstructured` func?

Reply:
no, I'm not asking you for applying the change, I just offered you the solution :wink:  your version is fine too :v: